sources/engine/Stride.Engine/Engine/InstancingUserArray.cs

```diff
@@ -0,0 +1,84 @@
+using System;
+using Stride.Core;
+using Stride.Core.Mathematics;
+using Stride.Engine.Processors;
+
+namespace Stride.Engine
+{
+    [DataContract("InstancingUserArray")]
+    [Display("UserArray")]
+    public class InstancingUserArray : InstancingBase
+    {
+        /// <summary>
+        /// The instance transformation matrices.
+        /// </summary>
+        [DataMemberIgnore]
+        public Matrix[] WorldMatrices = new Matrix[0];
+
+        /// <summary>
+        /// The inverse instance transformation matrices, updated automatically by the <see cref="InstancingProcessor"/>.
+        /// </summary>
+        [DataMemberIgnore]
+        public Matrix[] WorldInverseMatrices = new Matrix[0];
+
+        /// <summary>
+        /// A flag indicating whether the inverse matrices and bounding box should be calculated this frame.
+        /// </summary>
+        bool matricesUpdated;
+
+        /// <summary>
+        /// Updates the world matrices.
+        /// </summary>
+        /// <param name="matrices">The matrices.</param>
+        /// <param name="instanceCount">The instance count. When set to -1 the lenght if the matrices array is used</param>
+        public void UpdateWorldMatrices(Matrix[] matrices, int instanceCount = -1)
+        {
+            WorldMatrices = matrices;
+
+            if (WorldMatrices != null)
+            {
+                InstanceCount = instanceCount < 0 ? WorldMatrices.Length : Math.Min(WorldMatrices.Length, instanceCount);
+            }
+            else
+            {
+                InstanceCount = 0;
+            }
+
+            matricesUpdated = true;
+        }
+
+
+        public override void Update()
+        {
+            base.Update();
+
+            if (matricesUpdated)
+            {
+                if (WorldMatrices != null)
+                {
+                    // Make sure inverse matrices array is big enough
+                    if (WorldInverseMatrices.Length < InstanceCount)
+                    {
+                        WorldInverseMatrices = new Matrix[InstanceCount];
+                    }
+
+                    // Invert matrices and update bounding box
+                    var bb = BoundingBox.Empty;
+                    for (int i = 0; i < InstanceCount; i++)
```

Comment:
I could be wrong, but my gut tells me you should hold a local copy of `InstanceCount` in this method because it's a virtual property otherwise it'll take a minor performance hit.

Reply:
Good point!